### PR TITLE
TorchScript

### PR DESCRIPTION
**Pull request by erksch**
Applied the required changes to #100.

- Added to the README that torchscript works starting from PyTorch 1.10.0
- Set testing PyTorch version to >=1.10.0

### CODE REVIEW

File: tests/test_crf.py [hunk 1]
```diff
@@ -467,3 +467,130 @@ def test_first_timestep_mask_is_not_all_on(self):
...
+        assert torch.equal(llh_script, llh), f"scripted crf forward output {llh_script} " \
+                                             f"not matching non-scripted forward output {llh}"
+
+        # Test scripted forward when running batch first mode
+        crf_bf = make_crf(batch_first=True)
+        # Copy parameter values from non-batch-first CRF; requires_grad must be False
+        # to avoid runtime error of in-place operation on a leaf variable
+        crf_bf.start_transitions.requires_grad_(False).copy_(crf.start_transitions)
+        crf_bf.end_transitions.requires_grad_(False).copy_(crf.end_transitions)
+        crf_bf.transitions.requires_grad_(False).copy_(crf.transitions)
```
Comment (kmkurn) [inline tests/test_crf.py:541]: Why do you need to copy the parameters? Same question for similar changes in `test_scripted_decode`.

File: torchcrf/__init__.py [hunk 2]
```diff
@@ -151,16 +153,18 @@ def _validate(
                 f'got {emissions.size(2)}')
 
         if tags is not None:
-            if emissions.shape[:2] != tags.shape:
+            if emissions.shape[0] != tags.shape[0] or emissions.shape[1] != tags.shape[1]:
```
Comment (kmkurn) [inline torchcrf/__init__.py:156]: Why is this changed? Same question for other similar changes below.

File: torchcrf/__init__.py [hunk 3]
```diff
@@ -313,17 +317,18 @@ def _viterbi_decode(self, emissions: torch.FloatTensor,
...
             # for the last timestep
             _, best_last_tag = score[idx].max(dim=0)
-            best_tags = [best_last_tag.item()]
+            best_tags: List[int] = []
+            best_tags.append(best_last_tag.item())
 
             # We trace back where the best last tag comes from, append that to our best tag
             # sequence, and trace it back again, and so on
-            for hist in reversed(history[:seq_ends[idx]]):
+            for hist in history[:seq_ends[idx]][::-1]:
```
Comment (kmkurn) [inline torchcrf/__init__.py:331]: Why is this changed?

File: requirements-test.txt [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # This only installs PyTorch with a specific CUDA version which may not be
 # compatible with yours. If so, install PyTorch with the correct CUDA version
 # as instructed on https://pytorch.org/get-started/locally/
-torch
+torch>=1.10.0
```
Comment (kmkurn) [inline requirements-test.txt:4]: The PyTorch version that gets installed for testing is specified in `.github/workflows/run_tests.yml`. You should change the version there rather than here. That said, I'm not comfortable with testing only for v1.10.0 just to accommodate torch scripting. I think the ideal option is to run all tests *except* `TestTorchScript` for v1.2, and run *only* `TestTorchScript` on v1.10.0. That way, all the other tests are still run for <v1.10.0, ensuring that the library still supports those earlier versions. Would you be interested in implementing this?

**Review by kmkurn — COMMENTED** [review]
Thank you for the PR! I have left comments and questions. I'd appreciate if you could have a look at them.

**Comment by leejuyuu** [comment]
Hello @erksch, I would like to see this feature merged as it seems to enable ONNX export. Are you still interested? If not, would you mind if I try to resolve the reviews of @kmkurn on top of this branch (possibly in another PR because of push rights)?

File: torchcrf/__init__.py (hunk 3 above)
Comment (leejuyuu) [reply torchcrf/__init__.py:331]: This is needed because torchscript JIT does not support reversed yet.
https://github.com/pytorch/pytorch/issues/31772

File: torchcrf/__init__.py (hunk 2 above)
Comment (leejuyuu) [reply torchcrf/__init__.py:156]: This line of change is not necessary, but the cast to tuple below will error.

File: tests/test_crf.py (hunk 1 above)
Comment (leejuyuu) [reply tests/test_crf.py:541]: I believe these were copied from the original test code. These copying statements are also present in the tests above, e.g. https://github.com/kmkurn/pytorch-crf/blob/c885fba8b3754eb20520166c75ae2a50ae99847d/tests/test_crf.py#L272-L276.
